game-core/src/main/java/games/strategy/triplea/odds/calculator/BattleCalculator.java

```diff
@@ -16,69 +16,27 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
 import javax.annotation.Nonnull;
 import lombok.Setter;
 
 class BattleCalculator implements IBattleCalculator {
   @Nonnull private final GameData gameData;
-  private GamePlayer attacker = null;
-  private GamePlayer defender = null;
-  private Territory location = null;
-  private Collection<Unit> attackingUnits = new ArrayList<>();
-  private Collection<Unit> defendingUnits = new ArrayList<>();
-  private Collection<Unit> bombardingUnits = new ArrayList<>();
-  private Collection<TerritoryEffect> territoryEffects = new ArrayList<>();
   @Setter private boolean keepOneAttackingLandUnit = false;
   @Setter private boolean amphibious = false;
   @Setter private int retreatAfterRound = -1;
   @Setter private int retreatAfterXUnitsLeft = -1;
-  private boolean retreatWhenOnlyAirLeft = false;
   @Setter private String attackerOrderOfLosses = null;
   @Setter private String defenderOrderOfLosses = null;
   private volatile boolean cancelled = false;
-  private volatile boolean isRunning = false;
+  private final AtomicBoolean isRunning = new AtomicBoolean(false);
 
   BattleCalculator(final GameData data, final boolean dataHasAlreadyBeenCloned) {
     gameData =
         Preconditions.checkNotNull(
             dataHasAlreadyBeenCloned ? data : GameDataUtils.cloneGameData(data, false));
   }
 
-  /** Calculates odds using the stored game data. */
-  private void setCalculateData(
-      final GamePlayer attacker,
-      final GamePlayer defender,
-      final Territory location,
-      final Collection<Unit> attacking,
-      final Collection<Unit> defending,
-      final Collection<Unit> bombarding,
-      final Collection<TerritoryEffect> territoryEffects,
-      final boolean retreatWhenOnlyAirLeft)
-      throws IllegalStateException {
-    if (isRunning) {
-      return;
-    }
```

Comment:
Once you inline this method you realize that this return statement leads to undefined behaviour.
It either results in an exception because the GameData has not been set properly, or it makes the calculator code run on old data without any exceptions, but with potentially bad results